src/geometry/GeoTrackView.i.hh

```diff
@@ -0,0 +1,185 @@
+          stateview.vgstate, stateview.vgmaxdepth, id))
+    , vgnext_(GeoTrackView::get_nav_state(
+          stateview.vgnext, stateview.vgmaxdepth, id))
+    , pos_(stateview.pos[id.get()])
+    , dir_(stateview.dir[id.get()])
+    , next_step_(stateview.next_step[id.get()])
+{
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct the state.
+ *
+ * Expensive. This function should only be called to initialize an event from a
+ * starting location and direction. Secondaries will initialize their states
+ * from a copy of the parent.
+ */
+CELER_FUNCTION GeoTrackView& GeoTrackView::operator=(const Initializer_t& init)
+{
+    // Initialize position/direction
+    pos_ = init.pos;
+    dir_ = init.dir;
+
+    // Set up current state and locate daughter volume.
+    vgstate_.Clear();
+    const vecgeom::VPlacedVolume* volume         = shared_.world_volume;
+    const bool                    contains_point = true;
+
+    // Note that LocateGlobalPoint sets vgstate. If `vgstate_` is outside
+    // (including possibly on the outside volume edge), the volume pointer it
+    // returns will be null
+    vecgeom::GlobalLocator::LocateGlobalPoint(
+        volume, detail::to_vector(pos_), vgstate_, contains_point);
+
+    // Set up next state
+    vgnext_.Clear();
+    next_step_ = celeritas::numeric_limits<real_type>::quiet_NaN();
+    return *this;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Find the distance to the next geometric boundary.
+ */
+CELER_FUNCTION void GeoTrackView::find_next_step()
+{
+    vecgeom::VNavigator const* navigator
+        = this->volume().GetLogicalVolume()->GetNavigator();
+    next_step_
+        = navigator->ComputeStepAndPropagatedState(detail::to_vector(pos_),
+                                                   detail::to_vector(dir_),
+                                                   vecgeom::kInfLength,
+                                                   vgstate_,
+                                                   vgnext_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Move to the next boundary.
+ */
+CELER_FUNCTION void GeoTrackView::move_next_step()
+{
+    vgstate_ = vgnext_;
+    vgnext_.Clear();
+
+    // Move the next step plus an extra fudge distance
+    axpy(next_step_, dir_, &pos_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the volume ID in the current cell.
+ */
+CELER_FUNCTION VolumeId GeoTrackView::volume_id() const
+{
+    REQUIRE(!this->is_outside());
+    return VolumeId{this->volume().id()};
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Return whether the track is inside or outside the valid geometry region.
+ */
+CELER_FUNCTION Boundary GeoTrackView::boundary() const
+{
+    return (vgstate_.IsOnBoundary() ? Boundary::Yes : Boundary::No);
+}
```

Comment:
Since the original boundary was quite different than the VecGeom concept for boundary, I think it is easier to make them closer now before we start writing other code based on it.
So the question is who will be ultimately responsible for dealing with the track status.  I don't have the whole picture in mind yet.

Reply:
Removed boundary definitions.